modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/toolbarbuttons/plugin.js

```diff
@@ -13,14 +13,246 @@
+	const HEADING_NONE = 'None';
+	const HEADING_ROW = 'Row';
+
 	const pluginName = 'toolbarbuttons';
 
 	if (CKEDITOR.plugins.get(pluginName)) {
 		return;
 	}
 
+	/**
+	 * Creates a new CKEDITOR.dom.element using the passed tag name.
+	 *
+	 * @instance
+	 * @memberof CKEDITOR.Table
+	 * @protected
+	 * @method _createElement
+	 * @param {String} name The tag name from which an element should be created
+	 * @param {Object} editor The CKEditor instance.
+	 * @return {CKEDITOR.dom.element} Instance of CKEDITOR DOM element class
+	 */
+	function _createElement(name, editor) {
+		return new CKEDITOR.dom.element(name, editor.document);
+	}
+
+	/**
+	 * Returns which heading style is set for the given table.
+	 *
+	 * @instance
+	 * @memberof CKEDITOR.Table
+	 * @method getHeading
+	 * @param {CKEDITOR.dom.element} table The table to gather the heading from. If null, it will be retrieved from the current selection.
+	 * @param {Object} editor The CKEditor instance.
+	 * @return {String} The heading of the table. Expected values are `CKEDITOR.Table.NONE`, `CKEDITOR.Table.ROW`, `CKEDITOR.Table.COL` and `CKEDITOR.Table.BOTH`.
+	 */
+	function getHeading(table, editor) {
+		table = table || getFromSelection(editor);
+
+		if (!table) {
+			return null;
+		}
+
+		const rowHeadingSettings = table.$.tHead !== null;
+
+		let colHeadingSettings = true;
+
+		// Check if all of the first cells in every row are TH
+
+		for (let row = 0; row < table.$.rows.length; row++) {
+
+			// If just one cell isn't a TH then it isn't a header column
+
+			const cell = table.$.rows[row].cells[0];
+
+			if (cell && cell.nodeName.toLowerCase() !== 'th') {
+				colHeadingSettings = false;
+				break;
+			}
+		}
+
+		let headingSettings = HEADING_NONE;
+
+		if (rowHeadingSettings) {
+			headingSettings = HEADING_ROW;
+		}
+
+		if (colHeadingSettings) {
+			headingSettings =
+				headingSettings === HEADING_ROW ? HEADING_BOTH : HEADING_COL;
+		}
+
+		return headingSettings;
+	}
+
+	/**
+	 * Retrieves a table from the current selection.
+	 *
+	 * @instance
+	 * @memberof CKEDITOR.Table
+	 * @method getFromSelection
+	 * @param {Object} editor The CKEditor instance.
+	 * @return {CKEDITOR.dom.element} The retrieved table or null if not found.
+	 */
+	function getFromSelection(editor) {
+		let table;
+		const selection = editor.getSelection();
+		const selected = selection.getSelectedElement();
+
+		if (selected && selected.is('table')) {
+			table = selected;
+		}
+		else {
+			const ranges = selection.getRanges();
+
+			if (ranges.length > 0) {
+
+				// Webkit could report the following range on cell selection (#4948):
+				// <table><tr><td>[&nbsp;</td></tr></table>]
+
+				/* istanbul ignore else */
+				if (CKEDITOR.env.webkit) {
+					ranges[0].shrink(CKEDITOR.NODE_ELEMENT);
+				}
+
+				table = editor
+					.elementPath(ranges[0].getCommonAncestor(true))
+					.contains('table', 1);
+			}
+		}
+
+		return table;
+	}
+
+	function setHeading(table, heading, editor) {
```

Comment:
[optional] Generally I really like when a function receives an object, and we deconstruct it right away. It's so much easier to maintain code if we don't need to be careful about order of arguments. An update will not crash usage somewhere where we may not see.

What I mean is:
```js
setHeading({table, heading, editor})
```

Reply:
Yes, I like this approach too but I remember while working on editor stuff some months ago a bug caused by the losing of prototype when using destructuring but I can give a try and see how it behave

Reply:
I've done this, it worked except for some `this._setHeading` calls where it don't work

Reply:
> Yes, I like this approach too but I remember while working on editor stuff some months ago a bug caused by the losing of prototype when using destructuring but I can give a try and see how it behave

Great observation! Yeah, if we are dealing with inheritance, this is not a good pattern. But, if it's just a utility function, like here, we're good to go.

> I've done this, it worked except for some this._setHeading calls where it don't work

I don't see any reason why it wouldn't work :thinking: There is no inheritance with `_setHeading`, it's a basically a utility function, that happens to be defined as an object property. We're doing this in a lot of places, for example in [linktoolbar](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js#L23).